### ISSUE TITLE
On Fedora "Workstation Edition", iconv library is not identified by meson

### ISSUE THREAD

**Issue by catull** [att/ast#308]
My Fedora 27 is a "Workstation edition".

```sh
dnf info libiconv
```

gives me this:
```log
Installed Packages
Name         : libiconv
Version      : 1.15
Release      : 1.fc27
Arch         : x86_64
Size         : 3.2 M
Source       : libiconv-1.15-1.fc27.src.rpm
Repo         : @System
From repo    : @commandline
Summary      : Character set conversion library
URL          : https://www.gnu.org/software/libiconv/
License      : LGPL v2+
Description  : This library provides an iconv() implementation, for use on systems
             : which don't have one, or whose implementation cannot convert from/to
             : Unicode.
```

But `meson build` displays:

```log
........
Library iconv found: NO
.........
```

The relevant lines in build/meson-logs/meson-log.txt are:
```log
Library m found: YES
Running compile:
Working directory:  /tmp/tmpc2kxvlc1
Command line:  cc /tmp/tmpc2kxvlc1/testfile.c -O0 -Wl,--start-group -liconv -Wl,--end-group -pipe -D_FILE_OFFSET_BITS=64 -o /tmp/tmpc2kxvlc1/output.exe

Code:
 int main(int argc, char **argv) { return 0; }
Compiler stdout:

Compiler stderr:
 /usr/bin/ld: cannot find -liconv
collect2: error: ld returned 1 exit status

Library iconv found: NO
```

Do you also experience this on Fedora ?
Do you recommend to use "Fedora Server" instead ?

**Comment by catull** [comment]
Interestingly, the installation is in /usr/libiconv, with include/, lib/ (aliased to lib64/) and share/ sub-folders.

Thus, with `test-iconv.c` being
```c
#include <iconv.h>

int main(int argc, char **argv) {

  iconv_t result = iconv_open ("en_US.iso88591", "en_US.UTF-8");
  return 0;
}
```

If I compile it with `gcc -I/usr/libiconv/include -L/usr/libiconv/lib test-iconv.c -o test-iconv.exe` the output is:
```log
/tmp/ccXCDK57.o: In function `main':
test-iconv.c:(.text+0x1a): undefined reference to `libiconv_open'
collect2: error: ld returned 1 exit status
```

Hm, I have seen this before.
This version does succeed:
`gcc -DLIBICONV_PLUG=1 -I/usr/libiconv/include -L/usr/libiconv/lib -o test-iconv test-iconv.c`.

**Comment by catull** [comment]
If I `dnf remove libiconv`, I get the same behaviour on a clean workspace.
`meson build` still reports
```log
Library iconv found: NO
```

Is this known ?

**Comment by catull** [comment]
With libiconv removed, I see a header file named /usr/include/iconv.h.
It probably was there before.

The test C file above still does NOT compile with `cc test-iconv.c -O0 -Wl,--start-group -liconv -Wl,--end-group -pipe -D_FILE_OFFSET_BITS=64 -o test-iconv.exe` outside of meson.

But this one does:
`cc test-iconv.c -O0 -pipe -D_FILE_OFFSET_BITS=64 -o test-iconv.exe`

I the middle part (`-Wl,--start-group -liconv -Wl,--end-group`) necessary ?

**Comment by catull** [comment]
I cannot find a single occurrence of `start-group` on a clean workspace.

My guess is it comes from `meson`, because that string appeared in a few build/meson-* files after `meson build` on Fedora.

I do not see it in a workspace on macOS after `meson build`.

**Comment by catull** [comment]
This commit [b3d048e93a55feaedadb1efae72de0b8871eefab](https://github.com/mesonbuild/meson/commit/b3d048e93a55feaedadb1efae72de0b8871eefab) to `meson` added `-Wl,--start/end-group` for `dynamic libraries`, too, to fix their own issue [2096](https://github.com/mesonbuild/meson/issues/2096).
Before that fix, only static libraries were affected.

On Fedora I have `meson` version 0.43, installed with `dnf install meson`.

**Comment by catull** [comment]
On Fedora, `iconv_open` and `iconv_close` are both implemented in /lib64/libc-2.26.so.
So, strictly speaking, trying to link with `-liconv` is not succeeding.

So, lastly, this one does link succesfully: remove the `-liconv`, keep the --start|stop-group:
`cc test-iconv.c -O0 -Wl,--start-group -Wl,--end-group -pipe -D_FILE_OFFSET_BITS=64 -o test-iconv.exe`.

Perhaps the detection of the`iconv` library on Fedora is bit more complicated.

**Comment by catull** [comment]
Adding these lines to src/lib/libast/meson.build, after line 219:
```meson
if cc.has_function('iconv_open', prefix: '#include <iconv.h>')
    libast_c_args += [ '-D_lib_iconv_open' ]
endif
```
displays:
```log
....
Library iconv found: NO
Checking for function "iconv_open": YES
Library catgets found: NO
....
```
on Fedora, and
```log
Library iconv found: YES
Checking for function "iconv_open": NO
Library catgets found: NO
```
on macOS.

**Comment by krader1961** [comment]
TL;DR but on GNU Linux systems the iconv family of functions are in libc. On BSD systems they are in libiconv. I don't have Fedora installed but on Ubuntu Meson fails to find libiconv because it doesn't exist. Which is expected.

**Comment by krader1961** [comment]
Closing since I think I've answered the question. Feel free to reopen, @catull, if that isn't the case or you still feel there is a problem.
